### ISSUE TITLE
How to send IP of device which needs to shaped sent in payload of API calls.

### ISSUE THREAD

**Issue by muthurajbharathi** [facebook/augmented-traffic-control#260]
Hi I just need a Clarification of using REST API's , i dont want to Use Django's WEB UI for shaping device, rather i want to do it at Backend and i will specify IP of device that i want to Shape. and I want to Consume All  the REST apis by specifying IP.

There are 
1: GET http://192.168.0.1:8000/api/v1/token/
2: POST http://192.168.0.1:8000/api/v1/auth/192.168.0.13/
3:POST  http://192.168.0.1:8000/api/v1/shape/ 
4: GET http://192.168.0.1:8000/api/v1/shape/ 
5: GET and POST api/v1/profiles/
6:DELETE http://192.168.0.1:8000/api/v1/shape/

I want to Make all these API request from some different server not directly from the device that i want to shape.

How can i call All these API's right from token generation and authenticating client and getting profiles, Deleting profiles and Shaping device by Specifing IP of Device that i want to shape.

i Tried sending IP in URL of POST  http://192.168.0.1:8000/api/v1/shape/10.0.0.50 and also tried sending in payload 
 {"value":"10.0.0.50","down":{"rate":"1024","delay":{"delay":"2","jitter":0,"correlation":0},"loss":{"percentage":0,"correlation":0},"reorder":{"percentage":0,"correlation":0,"gap":0},"corruption":{"percentage":0,"correlation":0},"iptables_options":[]},"up":{"rate":"256","delay":{"delay":"2","jitter":0,"correlation":0},"loss":{"percentage":0,"correlation":0},"reorder":{"percentage":0,"correlation":0,"gap":0},"corruption":{"percentage":0,"correlation":0},"iptables_options":[]}}

--Right now I am trying to shape a Android Device which is connected to Wireless router and router is under atcd server's network.
--From another server machine i want to ATCD Django rest apis specifying IP of Device that i want to shape.

Please provide me  clarification how can i call all  above mentioned API's directly from backend  server instead of using web ui. and i want to specify IP of device i need to shape/auth/create profiles/ delete profiles.

**Comment by zfjagann** [comment]
Our APIs are designed (by default) to prevent one device from shaping a different device without their knowledge. In order to do this and maintain appropriate security, we implemented the `/token/` and `/auth/` endpoints.

The workflow for authorizing a device is:

`GET /token/` from the device you want to shape.
`POST /auth/{ip}` from the other device
`POST /shape/{ip}` from the other device

If this two-step authorization is not what you want and you're okay with anyone being able to shape anyone else in your organization, you can run ATCD with the `--atcd-mode unsecure` argument.
When run in this mode, the `/token/` and `/auth/` endpoints aren't necessary and you can just call `POST /shape/{ip}`

**Comment by muthurajbharathi** [comment]
Yes I tried unsecure mode and shaped from a server using curl whose
internet source and atcd machine's source internet are same and I was able
to shape and delete profiles, I hope from backend directly I can call atcd
apis for

GET create
POST create also
GET shape  in unsecure..

I am planning not to serve atcd ui directly. instead I will have my custom
ui page from some other machine whose internet and atcd machine's source
internet are same. where in backend I will make my server code to
GET create
POST create also
GET shape
POST shape in unsecure
DELETE shape.

And I have two more questions

1: please share your insights what are the other effects of running in ATCD
in unsecure mode. Except bypassing Auth.

2: In the Default mode or Secure mode, can I shape other devices from a
server which is not under atcd controlling wifi or LAN  router but ATCD
machine's internet source and server from which I want to shape are same.

Let's say

Machine 1 connected to network whose SSID is office

Machine 2 connected to network whose SSID is office

Machine 3 is atcd and whose source of Internet is SSID office

And atcd creates another network family to which we  have connected mobiles
and we open atcd ui directly from mobile to shape traffic of itself.

Mobiles will have IP series in 10.0.0.12 after its connected to ATC d's
network family.

But Machine 1 and Machine 2 will have 192.168.1.12 , 192.168.1.13.

ATCD input or source  IP is 192.16.1.15
And it creates another network family which is of 10.0.0.12 series and
device's connected to it will have ip's of 10.0.0.12 .

Basically I m trying to shape any mobile device from Machine1 or Machine2
from backend. Instead opening atcd ui in each mobile page. And I will
access atcd API using url example
192.168.1.15:8000/API/shape

192.168.1.15:8000/API/token
192.168.1.15:8000/API/create
192.168.1.15:8000/API/auth

In each requests I want to specify IP of device I want to shape , how do I
do that in secure mode.
For shape API I can pass in payload with address keyname.

Thanks alot for reply and i appreciate your suggestions.

**Comment by chantra** [comment]
> 1: please share your insights what are the other effects of running in ATCD
in unsecure mode. Except bypassing Auth.

no other

> 2: In the Default mode or Secure mode, can I shape other devices from a
server which is not under atcd controlling wifi or LAN  router but ATCD
machine's internet source and server from which I want to shape are same.

yes you can, you need to whitelist the ips through `PROXY_IPS` in https://github.com/facebook/augmented-traffic-control/blob/master/atc/django-atc-api/atc_api/settings.py IIRC

**Comment by muthurajbharathi** [comment]
Thanks alot for your suggestion helping me out...

**Comment by chantra** [comment]
@muthurajbharathi is this working for you now?

**Comment by muthurajbharathi** [comment]
Adding Proxy IP i have not tried but using *unsecure* mode i was able to
make curl requests to ATCD machine from a machine whose source of internet
and ATCD's source of internet are same.

example
machine A ip 10.0.0.11
machine B(Atcd machine ) 10.0.0.50
all the devices which are under control of ATCD's another network can be
shaped from making curl request to machine B using ip 10.0.0.80


example
curl --globoff -XDELETE http://10.0.0.50:8000/api/v1/shape/
 --data-urlencode address="$1"

this above request made from 10.0.0.11 machine . it worked for shaping and
unshaping any devices under atcd machine's controlled network.

let's say devices will have ip series like 192.0.0.1 which is under ATCD's
controlled network. so Shaping and Unshaping using curl requests  from a
Machine A 10.0.0.11 to Machine B 10.0.0.50 worked. Have to try creating
profile also.

so we can have a custom web page on any machine but in backend we can
directly talk to ATCD machine's api to control any devices which are under
ATCD's controlled network  that is what my next step.

Please correct me if am wrong on my way.

thanks
muthu










I will try adding proxy ips in settings and let you know the results.

**Comment by chantra** [comment]
@muthurajbharathi 
Is this working for you then?

**Comment by muthurajbharathi** [comment]
Adding IP not yet tried but from curl I was able to clear from machine
which is not under atcd controlled network...

**Comment by chantra** [comment]
@muthurajbharathi ok. Well, this should get you up and running. I will close this issue, please open a new one if you are still encountering issues after updating the config.
